File: backend/app/api/profiles.py
"""
Profile endpoints — onboarding and profile management.

Routes:
  GET   /profiles/me                      — get own profile (authenticated)
  PATCH /profiles/me                      — update own profile
  POST  /profiles/me/background           — upload background image to S3
  POST  /profiles/me/avatar               — upload avatar image to S3
  GET   /profiles/{profile_id}            — public profile (no auth required; is_public must be true)
  GET   /profiles/{profile_id}/inventory  — public inventory for a profile (active for-sale/trade items)
"""

import uuid as uuid_module
from decimal import Decimal
from typing import Optional, List, Dict, Any

import boto3
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File, status
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.db.session import get_db, settings
from app.dependencies import get_current_profile
from app.models.catalog_v2 import CardV2, ExpansionV2
from app.models.inventory import Inventory
from app.models.profiles import Profile

router = APIRouter(tags=["profiles"])

_VALID_ROLES = {"vendor", "collector"}


class ProfileUpdate(BaseModel):
    display_name: Optional[str] = Field(None, min_length=1, max_length=50)
    role: Optional[str] = None
    bio: Optional[str] = None
    tcg_interests: Optional[List[str]] = None
    zip_code: Optional[str] = Field(None, pattern=r"^\d{5}$")
    avatar_url: Optional[str] = None
    onboarding_complete: Optional[bool] = None
    buying_rate: Optional[Decimal] = Field(None, ge=0, le=1)
    trade_rate: Optional[Decimal] = Field(None, ge=0, le=1)
    is_accounting_enabled: Optional[bool] = None


def _public_profile_response(profile: Profile) -> Dict[str, Any]:
    """Safe subset of profile fields — returned to unauthenticated visitors."""
    return {
        "id": profile.id,
        "role": profile.role,
        "display_name": profile.display_name,
        "bio": profile.bio,
        "tcg_interests": profile.tcg_interests,
        "avatar_url": profile.avatar_url,
        "background_url": profile.background_url,
        "buying_rate": profile.buying_rate,
        "trade_rate": profile.trade_rate,
        "is_public": profile.is_public,
    }


def _inventory_item_response(inv: Inventory, card: Optional[CardV2], expansion: Optional[ExpansionV2]) -> Dict[str, Any]:
    def _image_url(images: Any) -> Optional[str]:
        if not images or not isinstance(images, list):
            return None
        return images[0].get("small") or images[0].get("large")

    return {
        "id": inv.id,
        "card_v2_id": inv.card_v2_id,
        "card_name": card.name if card else None,
        "set_name": expansion.name if expansion else None,
        "series_name": expansion.series if expansion else None,
        "card_num": card.number if card else None,
        "rarity": card.rarity if card else None,
        "image_url": _image_url(card.images) if card else None,
        "condition_type": inv.condition_type,
        "condition_ungraded": inv.condition_ungraded,
        "grading_company": inv.grading_company,
        "grade": inv.grade,
        "grading_company_other": inv.grading_company_other,
        "quantity": inv.quantity,
        "asking_price": float(inv.asking_price) if inv.asking_price is not None else None,
        "is_for_sale": inv.is_for_sale,
        "is_for_trade": inv.is_for_trade,
        "notes": inv.notes,
        "photo_url": inv.photo_url,
    }


def _profile_response(profile: Profile) -> Dict[str, Any]:
    return {
        "id": profile.id,
        "role": profile.role,
        "display_name": profile.display_name,
        "bio": profile.bio,
        "tcg_interests": profile.tcg_interests,
        "onboarding_complete": profile.onboarding_complete,
        "zip_code": profile.zip_code,
        "avatar_url": profile.avatar_url,
        "background_url": profile.background_url,
        "buying_rate": profile.buying_rate,
        "trade_rate": profile.trade_rate,
        "is_accounting_enabled": profile.is_accounting_enabled,
    }


@router.get("/profiles/me")
def get_profile(
    profile: Profile = Depends(get_current_profile),
) -> Dict[str, Any]:
    """Return the authenticated user's profile."""
    return _profile_response(profile)


@router.patch("/profiles/me")
def update_profile(
    body: ProfileUpdate,
    profile: Profile = Depends(get_current_profile),
    db: Session = Depends(get_db),
) -> Dict[str, Any]:
    """Update the authenticated user's profile."""
    update_data = body.model_dump(exclude_none=True)
    if not update_data:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="No fields to update")

    if "role" in update_data and update_data["role"] not in _VALID_ROLES:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid role. Must be one of: {', '.join(sorted(_VALID_ROLES))}",
        )

    for key, value in update_data.items():
        setattr(profile, key, value)

    db.add(profile)
    db.commit()
    db.refresh(profile)
    return _profile_response(profile)


@router.post("/profiles/me/background")
def upload_background(
    image: UploadFile = File(...),
    profile: Profile = Depends(get_current_profile),
    db: Session = Depends(get_db),
) -> Dict[str, str]:
    """Upload background image to S3, persist URL to profiles.background_url, return it."""
    if not image.content_type or not image.content_type.startswith("image/"):
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="File must be an image")

    image_bytes = image.file.read()
    if len(image_bytes) > 10 * 1024 * 1024:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Background image must be under 10 MB")

    if not settings.aws_s3_bucket:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Image storage not configured",
        )

    s3 = boto3.client(
        "s3",
        region_name=settings.aws_region,
        aws_access_key_id=settings.aws_access_key_id,
        aws_secret_access_key=settings.aws_secret_access_key,
    )
    key = f"backgrounds/{profile.id}/{uuid_module.uuid4()}.jpg"
    s3.put_object(
        Bucket=settings.aws_s3_bucket,
        Key=key,
        Body=image_bytes,
        ContentType=image.content_type,
    )
    url = f"https://{settings.aws_s3_bucket}.s3.amazonaws.com/{key}"
    profile.background_url = url
    db.add(profile)
    db.commit()
    return {"background_url": url}


@router.post("/profiles/me/avatar")
def upload_avatar(
    image: UploadFile = File(...),
    profile: Profile = Depends(get_current_profile),
    db: Session = Depends(get_db),
) -> Dict[str, str]:
    """Upload avatar image to S3 and return the public URL."""
    if not image.content_type or not image.content_type.startswith("image/"):
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="File must be an image")

    image_bytes = image.file.read()
    if len(image_bytes) > 5 * 1024 * 1024:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Avatar must be under 5 MB")

    if not settings.aws_s3_bucket:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Avatar storage not configured",
        )

    s3 = boto3.client(
        "s3",
        region_name=settings.aws_region,
        aws_access_key_id=settings.aws_access_key_id,
        aws_secret_access_key=settings.aws_secret_access_key,
    )
    key = f"avatars/{profile.id}/{uuid_module.uuid4()}.jpg"
    s3.put_object(
        Bucket=settings.aws_s3_bucket,
        Key=key,
        Body=image_bytes,
        ContentType=image.content_type,
    )
    url = f"https://{settings.aws_s3_bucket}.s3.amazonaws.com/{key}"
    profile.avatar_url = url
    db.add(profile)
    db.commit()
    return {"avatar_url": url}


# ---------------------------------------------------------------------------
# Public profile endpoints (no auth required)
# NOTE: these routes must be declared AFTER /profiles/me to avoid shadowing it.
# ---------------------------------------------------------------------------

@router.get("/profiles/{profile_id}")
def get_public_profile(
    profile_id: str,
    db: Session = Depends(get_db),
) -> Dict[str, Any]:
    """Return a public profile. Returns 404 if not found or is_public=False."""
    profile = db.get(Profile, profile_id)
    if profile is None or not profile.is_public:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Profile not found")
    return _public_profile_response(profile)


@router.get("/profiles/{profile_id}/inventory")
def get_public_profile_inventory(
    profile_id: str,
    db: Session = Depends(get_db),
) -> List[Dict[str, Any]]:
    """Return active for-sale/for-trade inventory for a public profile."""
    profile = db.get(Profile, profile_id)
    if profile is None or not profile.is_public:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Profile not found")

    rows = (
        db.query(Inventory, CardV2, ExpansionV2)
        .outerjoin(CardV2, Inventory.card_v2_id == CardV2.id)
        .outerjoin(ExpansionV2, CardV2.expansion_id == ExpansionV2.id)
        .filter(
            Inventory.profile_id == profile_id,
            Inventory.status == "active",
            Inventory.deleted_at.is_(None),
            (Inventory.is_for_sale.is_(True)) | (Inventory.is_for_trade.is_(True)),
        )
        .all()
    )

    return [_inventory_item_response(inv, card, expansion) for inv, card, expansion in rows]

File: frontend/lib/api/profiles.ts
/**
 * API client functions for /api/v1/profiles endpoints.
 */

import { getAccessToken } from "@/lib/supabase";

const API = process.env.NEXT_PUBLIC_API_URL!;

async function authHeaders(): Promise<HeadersInit> {
  const token = await getAccessToken();
  return token ? { Authorization: `Bearer ${token}` } : {};
}

/** Public-safe subset of profile fields returned by GET /profiles/{id} */
export interface PublicProfileData {
  id: string;
  role: "vendor" | "collector";
  display_name: string | null;
  bio: string | null;
  tcg_interests: string[] | null;
  avatar_url: string | null;
  background_url: string | null;
  buying_rate: number | null;
  trade_rate: number | null;
  is_public: boolean;
}

export interface ProfileData {
  id: string;
  role: "vendor" | "collector";
  display_name: string | null;
  bio: string | null;
  tcg_interests: string[] | null;
  onboarding_complete: boolean;
  zip_code: string | null;
  avatar_url: string | null;
  background_url: string | null;
  buying_rate: number | null;
  trade_rate: number | null;
  is_accounting_enabled: boolean;
}

export interface ProfileUpdate {
  display_name?: string;
  role?: "vendor" | "collector";
  bio?: string;
  tcg_interests?: string[];
  zip_code?: string;
  avatar_url?: string;
  onboarding_complete?: boolean;
  buying_rate?: number;
  trade_rate?: number;
  is_accounting_enabled?: boolean;
}

export async function getProfile(): Promise<ProfileData> {
  const res = await fetch(`${API}/api/v1/profiles/me`, {
    headers: await authHeaders(),
  });
  if (!res.ok) throw new Error((await res.json()).detail ?? "Failed to load profile");
  return res.json();
}

export async function updateProfile(data: ProfileUpdate): Promise<ProfileData> {
  const res = await fetch(`${API}/api/v1/profiles/me`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json", ...(await authHeaders()) },
    body: JSON.stringify(data),
  });
  if (!res.ok) throw new Error((await res.json()).detail ?? "Update failed");
  return res.json();
}

export async function uploadBackground(file: File): Promise<{ background_url: string }> {
  const token = await getAccessToken();
  const formData = new FormData();
  formData.append("image", file);
  const res = await fetch(`${API}/api/v1/profiles/me/background`, {
    method: "POST",
    headers: token ? { Authorization: `Bearer ${token}` } : {},
    body: formData,
  });
  if (!res.ok) throw new Error((await res.json()).detail ?? "Upload failed");
  return res.json();
}

export async function getPublicProfile(profileId: string): Promise<PublicProfileData> {
  const res = await fetch(`${API}/api/v1/profiles/${profileId}`);
  if (!res.ok) throw new Error((await res.json()).detail ?? "Profile not found");
  return res.json();
}

export async function uploadAvatar(file: File): Promise<{ avatar_url: string }> {
  const token = await getAccessToken();
  const formData = new FormData();
  formData.append("image", file);
  const res = await fetch(`${API}/api/v1/profiles/me/avatar`, {
    method: "POST",
    headers: token ? { Authorization: `Bearer ${token}` } : {},
    body: formData,
  });
  if (!res.ok) throw new Error((await res.json()).detail ?? "Upload failed");
  return res.json();
}
